wit/couchbase/subdocument.wit

```diff
@@ -0,0 +1,253 @@
+    /// If not specified, the implementer *may* provide a default timeout.
+    timeout-ms: option<u64>,
+
+    /// How and whether to retry the operation
+    retry-strategy: option<retry-strategy>,
+
+    /// A known span to associate this lookup with
+    parent-span: option<request-span>,
+  }
+
+  /// Errors that occur during lookup for an individual subdocument path
+  variant subdocument-lookup-error {
+    /// The provided path does not exist in the document
+    path-does-not-exist(string),
+    /// The path already exists (ex. during an insert)
+    path-already-exist(string),
+    /// Type Conflict between the path in the document and path command
+    path-mismatch(string),
+    /// A document error
+    document-error(document-error),
+    /// A completely unexpected error
+    unexpected(string),
+  }
+
+  /// Perform a lookup of one or more specs in a given document
+  lookup: func(
+    id: document-id,
+    operations: list<lookup-operation>,
+    options: lookup-options,
+  ) -> result<list<result<document, subdocument-lookup-error>>, lookup-error>;
+}
+
+/// Perform sub-document mutation operations on Documents stored in a Couchbase cluster
+///
+/// Reference: https://docs.couchbase.com/go-sdk/current/howtos/subdocument-operations.html#mutating
+interface subdocument-mutate {
+  use types.{
+    document-id, request-span, document, durability-level, subdocument-path, document-error, retry-strategy
+  };
+
+  /// Semantics to use for document level actions during mutation
+  enum store-semantics {
+    /// Replace the document
+    replace,
+    /// Replace the document or create if it doesn't exist
+    upsert,
+    /// Create the document, failing if it doesn't exist
+    insert,
+  }
+
+  /// Options for performing an insert mutation operation
+  record insert-options {
+    /// Whether the path should be created
+    create-path: bool,
+    /// Whether the mutate refers to an extended attribute, rather than the document itself
+    xattr: option<bool>,
+  }
+
+  /// Options for performing an upsert mutation operation
+  record upsert-options {
+    /// Whether the path should be created
+    create-path: bool,
+    /// Whether the mutate refers to an extended attribute, rather than the document itself
+    xattr: option<bool>,
+  }
+
+  /// Options for performing an replace mutation operation
+  record replace-options {
+    /// Whether the mutate refers to an extended attribute, rather than the document itself
+    xattr: option<bool>,
+  }
+
+  /// Options for performing an remove mutation operation
+  record remove-options {
+    /// Whether the mutate refers to an extended attribute, rather than the document itself
+    xattr: option<bool>,
+  }
+
+  /// Options for performing an array append mutation operation
+  record array-append-options {
+    /// Whether the path should be created
+    create-path: bool,
+    /// Adds mutliple elements to the array
+    has-multiple: bool,
```

Comment:
To avoid repeating the same mistakes that were made for the SDKs, why not remove this option and create a separate spec `array-append-multiple`, that requires an array, each element being added to the targeted array ?

Consistency is a good thing, but being consistent with an API that has been recognized as a mistake is subject to discussion imho.

Reply:
What do you think about having `array-append` and other related operations always take a list? We can avoid adding some more operations and I don't think it's *too* much trouble to create one-element lists.

Definitely a stylistic choice though so I could go either way.

Reply:
I've updated the contract to use the always-a-list construction -- please let me know what you think (should be easy to read and inspect at least)

Reply:
> What do you think about having `array-append` and other related operations always take a list? We can avoid adding some more operations and I don't think it's _too_ much trouble to create one-element lists.
> 
> Always taking lists versus adding an `array-append-multi`/`array-append-multiple` is definitely a stylistic choice though so I could go either way.

The issue with always accepting a list, is that `array-add-unique` does not support _multi_. So it would make the API inconsistent.

Reply:
Ah good point - to keep the API a bit more consistent I'll go with the `array-append` & `array-append-multi`!

- [ ] switch to `*-multi` operations